### DISCUSSION TITLE
Traefik v3 + LetsEncrypt

### DISCUSSION THREAD

**Discussion by Aetherinox** [Mailu/Mailu#3501] [User support]
Has anyone been able to successfully implement LetsEncrypt and Traefik v3 with Mailu?

Currently, the only way I can make it work is by generating an SSL certificate manually using certbot, and then adding the following to `mailu.env`

```ini
TLS_FLAVOR=cert
TLS_CERT_FILENAME=fullchain1.pem
TLS_KEYPAIR_FILENAME=privkey1.pem
```

I've read that the Cert Dumper image shouldn't be used anymore (deprecated), but I haven't been able to find any way to set this up to where I don't have to renew certificates manually.

**Comment by kirbylink** [comment]
That's a good question.
The documentation of Mailu 2.0 said the following:

> Sets the ``TLS_FLAVOR`` to one of the following
> values:
> 
> - ``cert`` is the default and requires certificates to be setup manually;
> - ``letsencrypt`` will use the *Letsencrypt!* CA to generate automatic certificates;
> - ``mail`` is similar to ``cert`` except that TLS will only be served for
>   emails (IMAP and SMTP), not HTTP (use it behind reverse proxies);
> - ``mail-letsencrypt`` is similar to ``letsencrypt`` except that TLS will only be served for
>   emails (IMAP and SMTP), not HTTP (use it behind reverse proxies);
> - ``notls`` will disable TLS, this is not recommended except for testing.

The documentation for Mailu 2024-06 refers only to

> Sets the ``TLS_FLAVOR`` to one of the following
> values:
> 
> - ``cert`` is the default and requires certificates to be setup manually;
> - ``letsencrypt`` will use the *Letsencrypt!* CA to obtain certificates automatically;
> - ``notls`` will disable TLS, this is not recommended except for testing.

And for Traefik as reverse proxy in mailu.env:
```
REAL_IP_FROM=192.168.203.0/24
PROXY_PROTOCOL=25,443,465,993,995,4190
TRAEFIK_VERSION=v2
TLS_FLAVOR=letsencrypt
WEBROOT_REDIRECT=/sso/login
```

Since they removed the cert-dumper i assume they want that Mailu handles the certificate on it's own if you need transport encryption (which should be the standard).
But that makes no sense for me since Traefik is handling my certificates with Wildcard-certificate (e. g. *.example.org)

The documentation isn't really good in explaining how to handle in the new version the certificates when cert-dumper shouldn't be used anymore while Traefik is still handling the certificates.

Would be good if a developer could tell how to handle such situation.

**Reply by nextgens** [reply]
Some work is underway on #3440 to clarify the said documentation.

The basic idea is that you should configure traefik to proxy **everything** matching vhosts/fqdns handled by Mailu to front using **TCP proxying** (Traefik shouldn't do TLS termination, it should redirect the traffic based on what SNI extensions say). This won't prevent you from letting Traefik handle the rest of the TLS traffic using the wildcard certificate for other services, as you currently do. Please note that Mailu also requires the traffic on port 80 for the vhosts/fqdns it handles to be sent to front, so that it can acquire certificates from letsencrypt using the http-01 method.

We do provide an example of how to configure it, I am not sure how much clearer we can make it... if you do please contribute to the PR.

**Reply by kirbylink** [reply]
@nextgens 
Yeah, you are right. There should be no problem with wildcard certificates.
I looked into my current setup and i had already a separated certificate for the mail subdomain.

But, will it still work when i have to handle two domains (e. g. mail.domain-a.com and mail.domain-b.com)?
Currently Traefik creates the additional certificate for mailu:

```yaml
  front:
    labels:
      - "traefik.enable=true"
      - ...
      - "traefik.http.routers.mailu.rule=Host(`mail.${HOSTNAME_EXTERN_DOMAIN_A}`) || Host(`mail.${HOSTNAME_EXTERN_DOMAIN_B}`)"
      - "traefik.http.routers.mailu.tls=true"
      - "traefik.http.routers.mailu.entrypoints=https"
      - "traefik.http.routers.mailu.tls.certresolver=letsencrypt-combination"
      - "traefik.http.routers.mailu.service=mailu"
```
Where `HOSTNAME_EXTERN_DOMAIN_A` and `HOSTNAME_EXTERN_DOMAIN_B` are two different domains?

In `mailu.env` i had:
```text
# Hostnames for this server, separated with comas
HOSTNAMES=mail.domain-a.com,mail.domain-b.com
```

**Reply by nextgens** [reply]
Yes it will work. Drop the existing labels and configure it as documented (without TLS on Traefik for Mailu's domains)
